Delete merged and squashed from other branches.

colonelpopcorn opened the pull request:
This defaults to master currently, my change proposes that the script should take an argument and allow squashed branches to be pruned no matter where they were merged.

colonelpopcorn committed:
Delete merged and squashed from other branches.

This defaults to master currently, my change proposes that the script should take an argument and allow squashed branches to be pruned no matter where they were merged.

amitbeck commented on bin/git-delete-squashed.js:
```diff
@@ -25,23 +26,27 @@ function git (args) {
 
+if (process.argv.length > 0) {
+  selectedBranchName = process.argv[1];
+}
```
Instead of `selectedBranchName || DEFAULT_BRANCH_NAME` everywhere, you could've just

```js
const selectedBranchName = process.argv[1] ?? DEFAULT_BRANCH_NAME;
```

or

```js
} else {
  selectedBranchName = DEFAULT_BRANCH_NAME;
}
```

colonelpopcorn replied:
I'm on Windows so that changes the order of arguments because I have to run the script with node.exe. Let me test on Mac and Linux to see what my updates will do.

Edit: Mac has the script as first argument (0th) as well. Will test on Linux to confirm my indexing is correct.

Edit Deux: Linux behaves as Mac and Windows do. This PR is good for review!

colonelpopcorn committed:
Incorporate feedback on passing in a branch name.

colonelpopcorn committed:
Add argument check to index properly.

On Windows the shebang/hashbang is not honored, we need some way to determine what order the arguments are in. Instead of grabbing an argument parsing library (still a possibility) we can just check to see if node has been used to run the script. Not ideal, but it is the best option I can see right now.

colonelpopcorn committed:
Fix eslint.

aladdin-add reviewed (commented):
you need to update the README too. :)

colonelpopcorn committed:
Update readme.

colonelpopcorn commented:
@aladdin-add Just updated it.

not-an-aardvark commented on README.md:
````diff
@@ -23,6 +23,14 @@ $ npm install --global git-delete-squashed
 $ git-delete-squashed
 ```
 
+You can specify an alternate branch to check for squashed merges, as well. This is useful for different names of trunk branches like `main` or `develop`.
````
Thanks for the pull request! Can you also update the bash script in the readme so that it also accepts an argument? It would be nice to keep the two versions of the script in sync.

not-an-aardvark replied:
(To clarify, I meant the script on line 14 of the readme, which currently hardcodes `master` rather than accepting an argument.)

colonelpopcorn replied:
Ok, that will require a fundamental change in how the one-liner is called. Are you okay with that?

EDIT: I say that because the only way the one liner can take an argument without a fundamental change is if evaluation/environment variables are used. I'm thinking I can change it to a function that defaults to `master` but can take an argument to override that.

colonelpopcorn replied:
@not-an-aardvark Just updated the readme.

colonelpopcorn committed:
Keep scripts together.

colonelpopcorn committed:
Add optional branch name to bash script.

not-an-aardvark commented on README.md:
````diff
@@ -11,16 +11,29 @@ This is useful if you work on a project that squashes branches into master. Afte
 ```bash
-git checkout -q master && git for-each-ref refs/heads/ "--format=%(refname:short)" | while read branch; do mergeBase=$(git merge-base master $branch) && [[ $(git cherry master $(git commit-tree $(git rev-parse $branch\^{tree}) -p $mergeBase -m _)) == "-"* ]] && git branch -D $branch; done
+# Change $TARGET_BRANCH to your targeted branch, e.g. change from `master` to `main` to delete branches squased into `main`.
+export TARGET_BRANCH=master && git checkout -q $TARGET_BRANCH && git for-each-ref refs/heads/ "--format=%(refname:short)" | while read branch; do mergeBase=$(git merge-base $TARGET_BRANCH $branch) && [[ $(git cherry $TARGET_BRANCH $(git commit-tree $(git rev-parse $branch\^{tree}) -p $mergeBase -m _)) == "-"* ]] && git branch -D $branch; done
````
```suggestion
TARGET_BRANCH=master && git checkout -q $TARGET_BRANCH && git for-each-ref refs/heads/ "--format=%(refname:short)" | while read branch; do mergeBase=$(git merge-base $TARGET_BRANCH $branch) && [[ $(git cherry $TARGET_BRANCH $(git commit-tree $(git rev-parse $branch\^{tree}) -p $mergeBase -m _)) == "-"* ]] && git branch -D $branch; done
```

I think this would avoid leaking the `TARGET_BRANCH` environment variable to subsequent commands.

not-an-aardvark commented on README.md:
```diff
@@ -11,16 +11,29 @@ This is useful if you work on a project that squashes branches into master. Afte
+# `git-delete-squashed` OR `git-delete-squased main`
+function git-delete-squashed() {
+    local targetBranch = ${$1-:master}
+    git checkout -q $targetBranch;
```
```suggestion
    git checkout -q $targetBranch &&
```

`&&` seems important here to avoid doing strange stuff and going off the rails if the branch doesn't exist. (I fixed this in the git alias recommendation above.)

colonelpopcorn replied:
I took your git alias suggestion, so this item is a non-issue now.

not-an-aardvark commented on README.md:
````diff
@@ -11,16 +11,29 @@ This is useful if you work on a project that squashes branches into master. Afte
 
 ```bash
-git checkout -q master && git for-each-ref refs/heads/ "--format=%(refname:short)" | while read branch; do mergeBase=$(git merge-base master $branch) && [[ $(git cherry master $(git commit-tree $(git rev-parse $branch\^{tree}) -p $mergeBase -m _)) == "-"* ]] && git branch -D $branch; done
+# Change $TARGET_BRANCH to your targeted branch, e.g. change from `master` to `main` to delete branches squased into `main`.
````
```suggestion
# Change $TARGET_BRANCH to your targeted branch, e.g. change from `master` to `main` to delete branches squashed into `main`.
```

not-an-aardvark commented on README.md:
```diff
@@ -11,16 +11,29 @@ This is useful if you work on a project that squashes branches into master. Afte
-git checkout -q master && git for-each-ref refs/heads/ "--format=%(refname:short)" | while read branch; do mergeBase=$(git merge-base master $branch) && [[ $(git cherry master $(git commit-tree $(git rev-parse $branch\^{tree}) -p $mergeBase -m _)) == "-"* ]] && git branch -D $branch; done
+# Change $TARGET_BRANCH to your targeted branch, e.g. change from `master` to `main` to delete branches squased into `main`.
+export TARGET_BRANCH=master && git checkout -q $TARGET_BRANCH && git for-each-ref refs/heads/ "--format=%(refname:short)" | while read branch; do mergeBase=$(git merge-base $TARGET_BRANCH $branch) && [[ $(git cherry $TARGET_BRANCH $(git commit-tree $(git rev-parse $branch\^{tree}) -p $mergeBase -m _)) == "-"* ]] && git branch -D $branch; done
+# OR you can put this function in your shell config and call it like this
```
I think it would be better to recommend a git alias:

```bash
$ git config --global alias.delete-squashed '!f() { local targetBranch=${1:-master} && git checkout -q $targetBranch && git branch --merged | grep -v "\*" | xargs -n 1 git branch -d && git for-each-ref refs/heads/ "--format=%(refname:short)" | while read branch; do mergeBase=$(git merge-base $targetBranch $branch) && [[ $(git cherry $targetBranch $(git commit-tree $(git rev-parse $branch^{tree}) -p $mergeBase -m _)) == "-"* ]] && git branch -D $branch; done; }; f'
```

Then it would be called as `git delete-squashed`.

colonelpopcorn replied:
This is a good suggestion. I was unaware of how powerful git aliases could be. Thanks for sharing your knowledge!

not-an-aardvark commented on README.md:
```diff
@@ -11,16 +11,29 @@ This is useful if you work on a project that squashes branches into master. Afte
+# OR you can put this function in your shell config and call it like this
+# `git-delete-squashed` OR `git-delete-squased main`
+function git-delete-squashed() {
+    local targetBranch = ${$1-:master}
```
I don't think this syntax is correct -- it should be `${1:-master}`. (I fixed this in the git alias recommendation above.)

not-an-aardvark reviewed (changes requested):
Thanks, and sorry about the long delay again. Just a couple suggestions.

colonelpopcorn committed:
Update README.md to not export target branch.

Co-authored-by: Teddy Katz <not-an-aardvark@users.noreply.github.com>

colonelpopcorn committed:
Fix typo.

Co-authored-by: Teddy Katz <not-an-aardvark@users.noreply.github.com>

colonelpopcorn committed:
Use git alias instead of bash function.
